rust/otap-dataflow/crates/admin/src/telemetry.rs

```diff
@@ -0,0 +1,866 @@
+        .attrs
+        .as_deref()
+        .unwrap_or("")
+        .split(',')
+        .map(|s| s.trim())
+        .filter(|s| !s.is_empty())
+        .collect();
+
+    // Aggregate groups with or without reset
+    let groups = aggregate_metric_groups(&state.metrics_registry, q.reset, &group_attrs);
+
+    // Resolve format (default json)
+    let fmt = q.format.unwrap_or_default();
+
+    match fmt {
+        OutputFormat::Json => {
+            let response = MetricsWithMetadata {
+                timestamp,
+                metric_sets: groups_with_metadata(&groups),
+            };
+            Ok(Json(response).into_response())
+        }
+        OutputFormat::JsonCompact => {
+            let response = AllMetrics {
+                timestamp,
+                metric_sets: groups_without_metadata(&groups),
+            };
+            Ok(Json(response).into_response())
+        }
+        OutputFormat::LineProtocol => {
+            let body = agg_line_protocol_text(&groups, Some(now.timestamp_millis()));
+            let mut resp = body.into_response();
+            let _ = resp.headers_mut().insert(
+                header::CONTENT_TYPE,
+                header::HeaderValue::from_static("text/plain; charset=utf-8"),
+            );
+            Ok(resp)
+        }
+        OutputFormat::Prometheus => {
+            let body = agg_prometheus_text(&groups, Some(now.timestamp_millis()));
+            let mut resp = body.into_response();
+            let _ = resp.headers_mut().insert(
+                header::CONTENT_TYPE,
+                header::HeaderValue::from_static("text/plain; version=0.0.4; charset=utf-8"),
+            );
+            Ok(resp)
+        }
+    }
+}
+
+fn aggregate_metric_groups(
+    registry: &MetricsRegistryHandle,
+    reset: bool,
+    group_attrs: &[&str],
+) -> Vec<AggregateGroup> {
+    // Aggregation map keyed by (set name, sorted list of (attr, Option<val_string>))
+    type GroupKey = (String, Vec<(String, Option<String>)>);
+    let mut agg: HashMap<
+        GroupKey,
+        (
+            HashMap<String, AttributeValue>,
+            HashMap<String, u64>,
+            &'static MetricsDescriptor,
+        ),
+    > = HashMap::new();
+
+    let mut visit = |descriptor: &'static MetricsDescriptor,
+                     attributes: &dyn AttributeSetHandler,
+                     metrics_iter: MetricsIterator<'_>| {
+        // Build key attributes vector
+        let mut key_attrs: Vec<(String, Option<String>)> = Vec::new();
+        if !group_attrs.is_empty() {
+            let mut amap: HashMap<&str, String> = HashMap::new();
+            for (k, v) in attributes.iter_attributes() {
+                let _ = amap.insert(k, v.to_string_value());
+            }
+            for gk in group_attrs {
+                let val_opt = amap.get(gk).cloned();
+                key_attrs.push((gk.to_string(), val_opt));
+            }
+        }
+        key_attrs.sort_by(|a, b| a.0.cmp(&b.0));
```

Comment:
Could we use unstable sorting here and other places in the code where we call `sort_by`?

```suggestion
        key_attrs.sort_unstable_by(|a, b| a.0.cmp(&b.0));
```